### PR TITLE
fix(sharing): [#FOR-636] improve sharing rights system

### PR DESCRIPTION
**Pull request by qperie**
## Describe your changes
We improved API security. Before, some API routes were secured by an AccessRight.class file only checking if the connected user had a workflow right to access the Formulaire mod.
That means anyone acessing Formulaire could call these API and get the responses.
We changed that by using the sharing security system : now, you need to have at least one sharing right on the form to call these API and get the infos.

To sum up :
- Before if acessing Formulaire you could get many infos of forms not related to you at all
- Now you need these forms to be shared to you as contributor, manager or responder at least to get these infos

All the refacto/improvement/modeling will be in the following ticket : https://jira.support-ent.fr/browse/FOR-692

## Issue ticket number and link
FOR-636 : https://jira.support-ent.fr/browse/FOR-636

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [ ] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: common/src/main/java/fr/openent/form/core/models/ShareObject.java [hunk 1]
```diff
@@ -80,27 +80,20 @@ public ShareObject model(JsonObject shareObject){
         return new ShareObject(shareObject);
     }
 
-    public ShareObject addCommonRights() {
-        JsonArray commonRights = RightsHelper.getCommonRights();
-
+    public void addCommonRights(JsonArray commonRights) {
         addCommonRightByField(USERS, commonRights);
         addCommonRightByField(GROUPS, commonRights);
         addCommonRightByField(BOOKMARKS, commonRights);
```
Comment (BastienBenard) [inline common/src/main/java/fr/openent/form/core/models/ShareObject.java:86]: A voir si on ne peut pas passer ces champs en model.

File: formulaire/src/main/java/fr/openent/formulaire/controllers/SharingController.java [hunk 2]
```diff
@@ -73,29 +74,29 @@ public void shareJson(final HttpServerRequest request) {
...
                     if (shareInfosEvt.isLeft()) {
-                        log.error("[Formulaire@SharingController::shareJson] Fail to get sharing infos : " + shareInfosEvt.left().getValue());
-                        renderInternalError(request, shareInfosEvt);
+                        log.error("[Formulaire@SharingController::shareResource] Fail to get sharing infos : " + shareInfosEvt.left().getValue());
+                        renderError(request);
                         return;
                     }
 
                     JsonObject shareInfos = shareInfosEvt.right().getValue().copy();
+                    JsonArray actions = shareInfos.getJsonArray(ACTIONS);
```
Comment (BastienBenard) [inline formulaire/src/main/java/fr/openent/formulaire/controllers/SharingController.java:83]: Je ne sais pas si tu peux pas utiliser des model en rapport au commentaire de common/src/main/java/fr/openent/form/core/models/ShareObject.java

File: formulaire/src/main/java/fr/openent/formulaire/controllers/SharingController.java [hunk 3]
```diff
@@ -73,29 +74,29 @@ public void shareJson(final HttpServerRequest request) {
...
-                            for (int i = actions.size() - 1; i >= 0; i--) {
-                                JsonObject action = actions.getJsonObject(i);
-                                if (action.getString(PARAM_DISPLAY_NAME).equals(RESPONDER_RESOURCE_RIGHT)) {
-                                    actions.remove(i);
-                                }
-                            }
+                        for (int i = actions.size() - 1; i >= 0; i--) {
+                            JsonObject action = actions.getJsonObject(i);
+                            boolean isReadRight = action.getString(PARAM_DISPLAY_NAME).equals(READ_RESOURCE_RIGHT);
+                            boolean isResponderRight = action.getString(PARAM_DISPLAY_NAME).equals(RESPONDER_RESOURCE_RIGHT);
```
Comment (BastienBenard) [inline formulaire/src/main/java/fr/openent/formulaire/controllers/SharingController.java:98]: Au cas au tu a un null inverse ton equals 
READ_RESOURCE_RIGHT.equals(action.getString(PARAM_DISPLAY_NAME))

File: formulaire/src/main/java/fr/openent/formulaire/controllers/SharingController.java [hunk 4]
```diff
@@ -146,78 +148,95 @@ public void shareResource(final HttpServerRequest request) {
...
-                        JsonArray responders = new JsonArray();
-                        for (int i = 0; i < infos.size(); i++) {
-                            JsonArray users = infos.getJsonObject(i).getJsonArray(USERS);
-                            responders.addAll(users);
+                        JsonArray ids = bookmarksEvt.right().getValue().getJsonObject(0).getJsonArray(IDS).getJsonObject(0).getJsonArray(IDS);
+                        for (int i = 0; i < ids.size(); i++) {
+                            JsonObject id = ids.getJsonObject(i);
+                            boolean isGroup = id.getString(NAME) != null;
+                            (isGroup ? groupsIds : usersIds).add(id.getString(ID));
                         }
```
Comment (BastienBenard) [inline formulaire/src/main/java/fr/openent/formulaire/controllers/SharingController.java:190]: Utilisation d'un stream

File: formulaire/src/main/java/fr/openent/formulaire/controllers/SharingController.java [hunk 5]
```diff
@@ -146,78 +148,95 @@ public void shareResource(final HttpServerRequest request) {
...
+                    Map<String, Object> idUsers = shareObject.getUsers().getMap();
+                    Map<String, Object> idGroups = shareObject.getGroups().getMap();
+                    Map<String, Object> idBookmarks = shareObject.getBookmarks().getMap();
+
+                    JsonArray usersIds = new JsonArray(new ArrayList<>(filterIdsForSending(idUsers).keySet()));
+                    JsonArray groupsIds = new JsonArray(new ArrayList<>(filterIdsForSending(idGroups).keySet()));
+                    JsonArray bookmarksIds = new JsonArray(new ArrayList<>(filterIdsForSending(idBookmarks).keySet()));
+
+                    // Get group ids and users ids from bookmarks and add them to previous lists
+                    neoService.getIdsFromBookMarks(bookmarksIds, bookmarksEvt -> {
```
Comment (BastienBenard) [inline formulaire/src/main/java/fr/openent/formulaire/controllers/SharingController.java:178]: tu peux polymorpher cette méthode en future et passer la méthode handler en deprecated

File: formulaire/src/main/java/fr/openent/formulaire/controllers/SharingController.java [hunk 6]
```diff
@@ -146,78 +148,95 @@ public void shareResource(final HttpServerRequest request) {
...
-                            if (syncEvt.isLeft()) {
-                                log.error("[Formulaire@SharingController::shareResource] " +
-                                        "Fail to sync distributions for form " + formId);
-                                renderInternalError(request, syncEvt);
+                            JsonArray infos = usersEvt.right().getValue();
+                            JsonArray responders = new JsonArray();
+                            for (int i = 0; i < infos.size(); i++) {
+                                JsonArray users = infos.getJsonObject(i).getJsonArray(USERS);
+                                responders.addAll(users);
+                            }
```
Comment (BastienBenard) [inline formulaire/src/main/java/fr/openent/formulaire/controllers/SharingController.java:206]: Utilisation de stream

File: formulaire/src/main/java/fr/openent/formulaire/controllers/SharingController.java [hunk 7]
```diff
@@ -146,78 +148,95 @@ public void shareResource(final HttpServerRequest request) {
...
                             }
 
-                            // Update 'collab' property as needed
-                            List<Map<String, Object>> idsObjects = new ArrayList<>();
-                            idsObjects.add(idUsers);
-                            idsObjects.add(idGroups);
-                            idsObjects.add(idBookmarks);
-                            updateFormCollabProp(formId, user, idsObjects, updateEvt -> {
-                                if (updateEvt.isLeft()) {
+                            syncDistributions(request, formId, user, responders, syncEvt -> {
```
Comment (BastienBenard) [inline formulaire/src/main/java/fr/openent/formulaire/controllers/SharingController.java:218]: tu peux polymorpher cette méthode en future et passer la méthode handler en deprecated

File: formulaire/src/main/java/fr/openent/formulaire/controllers/SharingController.java [hunk 8]
```diff
@@ -146,78 +148,95 @@ public void shareResource(final HttpServerRequest request) {
...
                                     return;
                                 }
-                                fixBugAutoUnsharing(request, formId, user, shareObject);
+
+                                // Update 'collab' property as needed
+                                List<Map<String, Object>> idsObjects = new ArrayList<>();
+                                idsObjects.add(idUsers);
+                                idsObjects.add(idGroups);
+                                idsObjects.add(idBookmarks);
+                                updateFormCollabProp(formId, user, idsObjects, updateEvt -> {
```
Comment (BastienBenard) [inline formulaire/src/main/java/fr/openent/formulaire/controllers/SharingController.java:231]: tu peux polymorpher cette méthode en future et passer la méthode handler en deprecated

**Comment by qperie** [comment]
@BastienBenard C'est tout noté pour le ticket refacto où tous ces aspects d'optimisation seront inclus

**Review by JinDouang — COMMENTED** [review]
ok pour refacto dans un autre ticket

**Pull request merged**